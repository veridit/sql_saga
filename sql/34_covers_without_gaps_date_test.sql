\i sql/include/test_setup.sql

BEGIN;

SET ROLE TO sql_saga_unprivileged_user;

CREATE TABLE date_shifts (
  job_id INTEGER,
  worker_id INTEGER,
  valid_from DATE,
  valid_until DATE
);
-- Add era and unique key via sql_saga
SELECT sql_saga.add_era('date_shifts', 'valid_from', 'valid_until', p_add_bounds_check := false);
SELECT sql_saga.add_unique_key('date_shifts', ARRAY['job_id', 'worker_id']);

TABLE sql_saga.era;
TABLE sql_saga.unique_keys;


-- Insert test data into the integer date_shifts table
INSERT INTO date_shifts(job_id, worker_id, valid_from, valid_until) VALUES
  (1, 1, '2017-11-23', '2017-11-25'),
  (1, 2, '2017-11-25', '2017-11-29'),
  (2, 3, '2017-11-23', '2017-11-25'),
  (2, 4, '2017-11-26', '2017-11-29'),
  (3, 5, '-infinity',  '2017-11-25'),
  (3, 5, '2017-11-25', '2017-11-29'),
  (4, 6, '2017-11-23', '2017-11-25'),
  (4, 7, '2017-11-25', 'infinity')
;

-- TRUE:

-- it covers when the range matches one exactly:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-25') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;

-- it covers when the range matches two exactly:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;

-- it covers when the range has extra in front:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-24', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;

-- it covers when the range has extra behind:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-27') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;

-- it covers when the range has extra on both sides:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-24', '2017-11-27') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;

-- an infinite start will cover a finite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 3;

-- an infinite start will cover an infinite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 3;

-- an infinite end will cover a finite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 4;

-- an infinite end will cover an infinite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', 'infinity') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 4;

-- FALSE:

-- it does not cover when the range is null:
SELECT  sql_saga.covers_without_gaps(NULL, daterange('2017-11-23', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range misses completely:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-30', '2017-12-01'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range has something at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range has something at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range has something in the middle:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 2;

-- it does not cover when the range is lower-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range is upper-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', 'infinity'))
FROM    date_shifts
WHERE   job_id = 1;

-- it does not cover when the range is both-sides-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', 'infinity'))
FROM    date_shifts
WHERE   job_id = 1;

-- an infinite start will not cover a finite target if there is uncovered time at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 3;

-- an infinite start will not cover an infinite target if there is uncovered time at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 3;

-- an infinite end will not cover a finite target if there is uncovered time at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 4;

-- an infinite end will not cover an infinite target if there is uncovered time at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', 'infinity'))
FROM    date_shifts
WHERE   job_id = 4;


-- NULL:

-- it is unknown when the target is null:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), null)
FROM    date_shifts
WHERE   job_id = 1;

-- Errors:

-- it fails if the input ranges go backwards:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until) , daterange('2017-11-26', '2017-11-28') ORDER BY worker_id DESC)
FROM    date_shifts
WHERE   job_id = 1;

-- TODO: handle an empty target range? e.g. [5, 5)
-- Or maybe since that is a self-contradiction maybe ignore that case?

DELETE FROM date_shifts;

SELECT sql_saga.drop_unique_key('date_shifts', ARRAY['job_id', 'worker_id'], 'valid');
SELECT sql_saga.drop_era('date_shifts');

DROP TABLE date_shifts;

ROLLBACK;

\i sql/include/test_teardown.sql
