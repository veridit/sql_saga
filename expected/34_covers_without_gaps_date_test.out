\i sql/include/test_setup.sql
--
-- test_setup.sql
--
-- Common setup for regression tests that need to be self-contained.
-- This script creates the extension, a user role, and grants permissions.
--
SET datestyle = 'ISO, YMD';
CREATE EXTENSION IF NOT EXISTS btree_gist;
CREATE EXTENSION IF NOT EXISTS sql_saga CASCADE;
DO $$
BEGIN
    CREATE ROLE sql_saga_unprivileged_user;
EXCEPTION WHEN duplicate_object THEN
END
$$;
GRANT USAGE ON SCHEMA sql_saga TO sql_saga_unprivileged_user;
GRANT SELECT ON ALL TABLES IN SCHEMA sql_saga TO sql_saga_unprivileged_user;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA sql_saga TO sql_saga_unprivileged_user;
/*
 * Allow the unprivileged user to create tables in the public schema.
 * This is required for tests that create their own tables.
 * PG 15+ restricts this by default.
 */
GRANT CREATE ON SCHEMA public TO PUBLIC;
BEGIN;
SET ROLE TO sql_saga_unprivileged_user;
CREATE TABLE date_shifts (
  job_id INTEGER,
  worker_id INTEGER,
  valid_from DATE,
  valid_until DATE
);
-- Add era and unique key via sql_saga
SELECT sql_saga.add_era('date_shifts', 'valid_from', 'valid_until');
 add_era 
---------
 t
(1 row)

SELECT sql_saga.add_unique_key('date_shifts', ARRAY['job_id', 'worker_id']);
           add_unique_key           
------------------------------------
 date_shifts_job_id_worker_id_valid
(1 row)

TABLE sql_saga.era;
 table_schema | table_name  | era_name | valid_from_column_name | valid_until_column_name | range_type | bounds_check_constraint | audit_schema_name | audit_table_name 
--------------+-------------+----------+------------------------+-------------------------+------------+-------------------------+-------------------+------------------
 public       | date_shifts | valid    | valid_from             | valid_until             | daterange  | date_shifts_valid_check |                   | 
(1 row)

TABLE sql_saga.unique_keys;
          unique_key_name           | table_schema | table_name  |    column_names    | era_name |                    unique_constraint                    |             exclude_constraint              
------------------------------------+--------------+-------------+--------------------+----------+---------------------------------------------------------+---------------------------------------------
 date_shifts_job_id_worker_id_valid | public       | date_shifts | {job_id,worker_id} | valid    | date_shifts_job_id_worker_id_valid_from_valid_until_key | date_shifts_job_id_worker_id_daterange_excl
(1 row)

-- Insert test data into the integer date_shifts table
INSERT INTO date_shifts(job_id, worker_id, valid_from, valid_until) VALUES
  (1, 1, '2017-11-23', '2017-11-25'),
  (1, 2, '2017-11-25', '2017-11-29'),
  (2, 3, '2017-11-23', '2017-11-25'),
  (2, 4, '2017-11-26', '2017-11-29'),
  (3, 5, '-infinity',  '2017-11-25'),
  (3, 5, '2017-11-25', '2017-11-29'),
  (4, 6, '2017-11-23', '2017-11-25'),
  (4, 7, '2017-11-25', 'infinity')
;
-- TRUE:
-- it covers when the range matches one exactly:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-25') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- it covers when the range matches two exactly:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- it covers when the range has extra in front:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-24', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- it covers when the range has extra behind:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-27') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- it covers when the range has extra on both sides:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-24', '2017-11-27') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- an infinite start will cover a finite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 3;
 covers_without_gaps 
---------------------
 t
(1 row)

-- an infinite start will cover an infinite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 3;
 covers_without_gaps 
---------------------
 t
(1 row)

-- an infinite end will cover a finite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 4;
 covers_without_gaps 
---------------------
 t
(1 row)

-- an infinite end will cover an infinite target:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', 'infinity') ORDER BY valid_from)
FROM    date_shifts
WHERE   job_id = 4;
 covers_without_gaps 
---------------------
 t
(1 row)

-- FALSE:
-- it does not cover when the range is null:
SELECT  sql_saga.covers_without_gaps(NULL, daterange('2017-11-23', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range misses completely:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-30', '2017-12-01'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range has something at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range has something at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range has something in the middle:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 2;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range is lower-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range is upper-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', 'infinity'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- it does not cover when the range is both-sides-unbounded:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', 'infinity'))
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 f
(1 row)

-- an infinite start will not cover a finite target if there is uncovered time at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-23', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 3;
 covers_without_gaps 
---------------------
 f
(1 row)

-- an infinite start will not cover an infinite target if there is uncovered time at the end:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('-infinity', '2017-11-30'))
FROM    date_shifts
WHERE   job_id = 3;
 covers_without_gaps 
---------------------
 f
(1 row)

-- an infinite end will not cover a finite target if there is uncovered time at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', '2017-11-29'))
FROM    date_shifts
WHERE   job_id = 4;
 covers_without_gaps 
---------------------
 f
(1 row)

-- an infinite end will not cover an infinite target if there is uncovered time at the beginning:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), daterange('2017-11-22', 'infinity'))
FROM    date_shifts
WHERE   job_id = 4;
 covers_without_gaps 
---------------------
 f
(1 row)

-- NULL:
-- it is unknown when the target is null:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until), null)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 
(1 row)

-- Errors:
-- it fails if the input ranges go backwards:
SELECT  sql_saga.covers_without_gaps(daterange(valid_from,valid_until) , daterange('2017-11-26', '2017-11-28') ORDER BY worker_id DESC)
FROM    date_shifts
WHERE   job_id = 1;
 covers_without_gaps 
---------------------
 t
(1 row)

-- TODO: handle an empty target range? e.g. [5, 5)
-- Or maybe since that is a self-contradiction maybe ignore that case?
DELETE FROM date_shifts;
SELECT sql_saga.drop_unique_key('date_shifts', ARRAY['job_id', 'worker_id'], 'valid');
 drop_unique_key 
-----------------
 
(1 row)

SELECT sql_saga.drop_era('date_shifts');
 drop_era 
----------
 t
(1 row)

DROP TABLE date_shifts;
ROLLBACK;
\i sql/include/test_teardown.sql
--
-- test_teardown.sql
--
-- Common teardown for regression tests. This script drops the unprivileged
-- user role created by test_setup.sql.
--
-- It is important to reset the role first, in case a test fails and
-- leaves the session role set to the user that is about to be dropped.
RESET ROLE;
-- Drop the extensions to ensure a clean state for the next test.
-- Use CASCADE to remove any dependent objects created by sql_saga.
DROP EXTENSION IF EXISTS sql_saga CASCADE;
DROP EXTENSION IF EXISTS btree_gist CASCADE;
-- Revoke any privileges held by the test user and drop any objects they own.
-- This is necessary before the role can be dropped.
DROP OWNED BY sql_saga_unprivileged_user;
DROP ROLE IF EXISTS sql_saga_unprivileged_user;
